concepts/eventing-code-guidelines.md

```diff
@@ -0,0 +1,1155 @@
+		t.Fatalf("Subscriber did not receive the message with error: %v", err)
+	}
+}
+
+func TestSendCloudEventWithReconnect(t *testing.T) { 
+
+	logger := logrus.New() 
+	logger.Info("TestNatsSender started") 
+
+	// Start Nats server 
+	natsServer := testingutils.StartNatsServer() 
+	assert.NotNil(t, natsServer) 
+	defer natsServer.Shutdown() 
+	// connect to nats 
+	bc := pkgnats.NewBackendConnection(natsServer.ClientURL(), true, 10, time.Second) 
+	err := bc.Connect() 
+	assert.Nil(t, err) 
+	assert.NotNil(t, bc.Connection) 
+
+	// create message sender 
+	ctx := context.Background() 
+	sender := NewNatsMessageSender(ctx, bc, logger) 
+
+	// subscribe to subject 
+	done := make(chan bool, 1) 
+	validator := testingutils.ValidateNatsMessageDataOrFail(t, fmt.Sprintf(`"%s"`, testingutils.CloudEventData), done) 
+	testingutils.SubscribeToEventOrFail(t, bc.Connection, testingutils.CloudEventType, validator) 
+
+	// create cloudevent 
+	ce := cloudevents.NewEvent() 
+	ce.SetType(testingutils.CloudEventType) 
+	err = json.Unmarshal([]byte(testingutils.StructuredCloudEventPayloadWithCleanEventType), &ce) 
+	assert.Nil(t, err) 
+
+	sendEventAndAssertStatus(ctx, t, sender, &ce, http.StatusNoContent) 
+
+	// wait for subscriber to receive the messages 
+	if err := testingutils.WaitForChannelOrTimeout(done, time.Second*3); err != nil { 
+		t.Fatalf("Subscriber did not receive the message with error: %v", err) 
+	} 
+
+	// close connection 
+	bc.Connection.Close() 
+	assert.True(t, bc.Connection.IsClosed()) 
+	sendEventAndAssertStatus(ctx, t, sender, &ce, http.StatusNoContent) 
+} 
+``` 
+
+</details>
+
+<details>
+  <summary>Do</summary>
+
+```go 
+func TestSendCloudEventsToNats(t *testing.T) {
+	testCases := []struct { 
+		name                               string
+		givenRetries                       int
+		wantHTTPStatusCode                 int
+		wantReconnectAfterConnectionClosed bool
+	}{ 
+		{
+			name:                              "sending event to NATS works",
+			givenRetries:                      1,
+			wantHTTPStatusCode:                http.StatusNoContent,
+			wantClosedConnectionBeforeSending: false,
+		},
+		{
+			name:               "sending event to NATS works given a closed connection",
+			givenRetries:       10,
+			wantHTTPStatusCode: http.StatusNoContent,
+			// Close connection before sending so we can check the reconnect behaviour of the NATS connection.
+			wantClosedConnectionBeforeSending: true,
+		},
+	} 
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			testEnv := setupTestEnvironment(t, pkgnats.WithMaxReconnects(tc.givenRetries))
+
+			// subscribe to subject
+			subject := fmt.Sprintf(`"%s"`, testingutils.CloudEventData)
+			// NOTE: we are using the testEnv.natsRecvConnection instead of testEnv.natsSendConnection because the latter will get reconnected based on wantClosedConnectionBeforeSending, this will fail the test when trying to receive a message.
```

Comment:
```suggestion
			// NOTE: We are using the testEnv.natsRecvConnection instead of testEnv.natsSendConnection because the latter will get reconnected based on wantClosedConnectionBeforeSending. This will fail the test when trying to receive a message.
```